kobo/shortcuts.py

```diff
@@ -271,8 +271,16 @@ def run(cmd, show_cmd=False, stdout=False, logfile=None, can_fail=False, workdir
     @return: (command return code, merged stdout+stderr)
     @rtype: (int, str) or (int, None)
     """
+    log_fd = None
     if logfile:
         logfile = os.path.join(workdir or "", logfile)
+        # What happens to log file if it exists already? If show_cmd is True,
+        # it will be overwritten. Otherwise the command output will just be
+        # appended to the existing file.
+        flags = (os.O_RDWR | os.O_APPEND
+                 if not show_cmd and os.path.exists(logfile)
+                 else os.O_WRONLY | os.O_CREAT | os.O_TRUNC)
+        log_fd = os.open(logfile, flags, 0o644)
```

Comment:
Won't this leak if any exception is raised between now and when you close it?

I note that this is using os.open which returns a raw fd, not a file-like object with a destructor, so garbage collector won't help here.